WordPress/gutenberg
# Introduce distraction free mode

## What?


This PR is the result of the exploration in #38928 to introduce a 
distraction free mode. This mode build on top of the reduce interface 
preference, amplifying its effect.

## Why?


To enhance content production in WordPress. Text based content is one of 
the most important forms of content. While the site and page building 
enhancements make multi media content production much easier, through the 
visual tools offered by the block editor, text content production can 
sometimes feel impaired.

A distraction free mode (the term became somewhat of an industry standard) 
aims to remove visual clutter when the focus is on text based content 
production.

## How?


The PR builds on the reduce interface prefference. Until now the 
preference, when toggled on, would only hide some of the toolbar icons and 
a few other non-essential UI items. With this PR the prefference will have 
a more drastic effect, turning the UI into a blank page:

- hides the top toolbar, makes it appear on hover over its normal occupied 
area
- removes many top toolbar buttons
- automatically closes any open sidebars
- hides the insertion point indicator
- hides the block toolbar

## Known issues

- [x] There is a mystery bug that makes the page scroll to top when 
distraction free mode is on and the `ESC` key is pressed 
- [x] There are some failing tests
- [x] Move focus from the options menu to somewhere useful (like the options toggle)
- [x] On small viewports the setting should have no effect, but it does 
break some things
- - The mobile viewport size hides all the "View" group of options in the Editor Options Menu
- [x] The copy is still unsettled (distraction free, toggle interface ...)
- [x] The `hasReducedUi` props need to be refactored into `isDistractionFree`
- [x] Shift+Tab or Alt+F10  should show the toolbar
- [x] The system notifications can't be closed because toolbar is on top
- [x] The chrome should be animated off screen not simply disappear. 
- [x] Snackbar notifications should adjust their bottom position

## Testing Instructions





1. Create a new post
2. Go to top bar > tools menu and click "Toggle interface"
3. Observe serenity

## Screenshots or screencast 

https://user-images.githubusercontent.com/107534/189492067-f2b797fc-a363-47e1-bbac-c96b89d48a4e.mp4
